### PR TITLE
fix(H2 Client): bind stream 'data' listener only after received 'response' event

### PR DESCRIPTION
**Pull request by St3ffGv4**
## This relates to...

https://github.com/nodejs/undici/issues/2808

## Rationale

Sporadically when doing massive h2 requests for a certain period of time a breaking assertion is triggered "assert(!this.aborted)" in onHeaders Method [undici/lib/core/request.js] causing application crash

## Changes

 - lib/dispatcher/client-h2.js

### Features

N/A

### Bug Fixes

- Sometimes happens that H2 Stream "data" event is triggered some milliseconds before the "response" event, this leads to an uninitialized body. To avoid this breaking error if the onData is called before onHeaders, return false so the stream is paused in client-h2.js

## Status

- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [x] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md #

### CODE REVIEW

**Review by ronag — CHANGES_REQUESTED** [review]
This need to be fixed somewhere else. Can we defer the onData in the h2 wrapper?

**Comment by St3ffGv4** [comment]
@ronag like doing this?

in client-h2.js

```javascript
let isResponseStarted = false

stream.once('response', headers => {
    isResponseStarted = true;
    const { [HTTP2_HEADER_STATUS]: statusCode, ...realHeaders } = headers
    request.onResponseStarted()

    if (request.onHeaders(Number(statusCode), realHeaders, stream.resume.bind(stream), '') === false) {
      stream.pause()
    }
  })
  
  stream.on('data', (chunk) => {
    if (!isResponseStarted || request.onData(chunk) === false) {
      stream.pause()
    }
  })
```

**Review by mcollina — CHANGES_REQUESTED** [review]
Thanks for opening a PR! Can you please add a unit test?

**Comment by St3ffGv4** [comment]
@mcollina this would maybe be a long term test, because the error can happen at any time, immediately or not, it would be like using the script reported in the issue and wait indefinitely if the scenario occurs

File: lib/dispatcher/client-h2.js [hunk 1]
```diff
@@ -419,7 +421,7 @@ function writeH2 (client, request) {
   })
 
   stream.on('data', (chunk) => {
-    if (request.onData(chunk) === false) {
+    if (!isStreamResponseStarted || request.onData(chunk) === false) {
```
Comment (ronag) [inline lib/dispatcher/client-h2.js:424]: this will throw away the first chunk?

Comment (St3ffGv4) [reply lib/dispatcher/client-h2.js:424]: @ronag thaks, good point, I was focused on seeing that the error no longer occurs.

Is it ok to declare a temp chunks array and if isStreamResponseStarted is false in 'data' event push the chunk and then in 'response' event if the temp chunks array has length > 0 call the request.onData method after the request.onHeaders?

Comment (ronag) [reply lib/dispatcher/client-h2.js:424]: This does kind of feel like an error in node core http, it doesn't make sense to me that data is emitted before response. Do you think you could report it in the node core repo in addition to this workaround PR.

Comment (St3ffGv4) [reply lib/dispatcher/client-h2.js:424]: @ronag before report,
i was checking nodejs documentation about http2

there's one example here: https://nodejs.org/api/http2.html#clienthttp2sessionrequestheaders-options
```javascript
const http2 = require('node:http2');
const clientSession = http2.connect('https://localhost:1234');
const {
  HTTP2_HEADER_PATH,
  HTTP2_HEADER_STATUS,
} = http2.constants;

const req = clientSession.request({ [HTTP2_HEADER_PATH]: '/' });
req.on('response', (headers) => {
  console.log(headers[HTTP2_HEADER_STATUS]);
  req.on('data', (chunk) => { /* .. */ });
  req.on('end', () => { /* .. */ });
});
```

the data event listener is  binded inside the response listener callback.
So i have moved client-h2.js stream on data event inside stream on response (at the and of the callback) and it seems to be ok now

```javascript
 stream.once('response', headers => {
    const { [HTTP2_HEADER_STATUS]: statusCode, ...realHeaders } = headers
    request.onResponseStarted()

    if (request.onHeaders(Number(statusCode), realHeaders, stream.resume.bind(stream), '') === false) {
      stream.pause()
    }

    stream.on('data', (chunk) => {
      if (request.onData(chunk) === false) {
        stream.pause()
      }
    })
  })
``` 

my test script is:
```javascript
const Agent = require("./lib/dispatcher/agent");
const { fetch } = require("./lib/web/fetch");

const dispatcher = new Agent({ allowH2: true });

(async () => {
  while (true) {
    //console.log("fetching 1000 requests");
    //console.time();
    await Promise.all(
      Array.from({ length: 1000 }).map(async () => {
        try {
          const response = await fetch("https://http2.pro/api/v1", {
            dispatcher,
          });
          const text = await response.text();
          //console.log(text);
          if(!text) throw new Error("NOT GOOD")
        } catch (e) {
          console.error(e);
        }
      }),
    );
    //console.timeEnd();
  }
})();
``` 

This make more sense to you?
Thanks

Comment (ronag) [reply lib/dispatcher/client-h2.js:424]: Make sense, but yea, please open an issue because this seems wrong.

Comment (St3ffGv4) [reply lib/dispatcher/client-h2.js]: Issue opened

**Review by ronag — APPROVED** [review]

File: lib/dispatcher/client-h2.js [hunk 2]
```diff
@@ -394,6 +394,12 @@ function writeH2 (client, request) {
     if (request.onHeaders(Number(statusCode), realHeaders, stream.resume.bind(stream), '') === false) {
       stream.pause()
     }
+
+    stream.on('data', (chunk) => {
```
Comment (metcoder95) [inline lib/dispatcher/client-h2.js:398]: Can we add the link to the issue as comment?

**Review by metcoder95 — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**